wolfcrypt/src/evp.c

```diff
@@ -2211,19 +2248,54 @@ int wolfSSL_EVP_PKEY_copy_parameters(WOLFSSL_EVP_PKEY *to,
         }
         break;
 #endif
-#ifndef NO_RSA
-    case EVP_PKEY_RSA:
-#endif
 #ifndef NO_DH
     case EVP_PKEY_DH:
+        if (from->dh) {
+            WOLFSSL_BIGNUM* cpy;
+            if (!to->dh && !(to->dh = wolfSSL_DH_new())) {
+                WOLFSSL_MSG("wolfSSL_DH_new error");
+                return WOLFSSL_FAILURE;
+            }
+
+            /* free existing BIGNUMs if needed before copying over new */
+            wolfSSL_BN_free(to->dh->p);
+            wolfSSL_BN_free(to->dh->g);
+            wolfSSL_BN_free(to->dh->q);
+            to->dh->p = NULL;
+            to->dh->g = NULL;
+            to->dh->q = NULL;
+
+            if (!(cpy = wolfSSL_BN_dup(from->dh->p))) {
+                WOLFSSL_MSG("wolfSSL_BN_dup error, DH p");
+                return WOLFSSL_FAILURE;
+            }
+            to->dh->p = cpy;
+            if (!(cpy = wolfSSL_BN_dup(from->dh->g))) {
+                WOLFSSL_MSG("wolfSSL_BN_dup error, DH g");
```

Comment:
Will this leak memory since we aren't freeing the `p` we just duplicated?

Reply:
A few lines above this immediate copy section, I'm first freeing `to->dh->p/g/q` with `wolfSSL_BN_free()`.  I think that might be what you were looking for here?

Reply:
I'm talking about `to->dh->p`, which is allocated at line 2277 and isn't freed in this error case at line 2283.

Reply:
As long as the EVP_PKEY owns the WOLFSSL_DH struct, any `to->dh` members that are not NULL should get freed when the application calls `EVP_PKEY_free()`.  I do see that if we allocate a new WOLFSSL_DH, we need to set `to->ownDh = 1`, otherwise it wouldn't be freed during EVP_PKEY_free().  I'll make that switch, and adjust the DSA/ECC cases above.